Source/ASNetworkImageNode.mm

```diff
@@ -583,24 +596,55 @@ - (void)_downloadImageWithCompletion:(void (^)(id <ASImageContainerProtocol> ima
     // We need to reobtain the lock after and ensure that the task we've kicked off still matches our URL. If not, we need to cancel
     // it and try again.
     {
       ASLockScopeSelf();
       url = self->_URL;
+      interfaceState = self->_interfaceState;
     }
 
-
-    downloadIdentifier = [self->_downloader downloadImageWithURL:url
-                                             callbackQueue:[self callbackQueue]
-                                          downloadProgress:NULL
-                                                completion:^(id <ASImageContainerProtocol> _Nullable imageContainer, NSError * _Nullable error, id  _Nullable downloadIdentifier, id _Nullable userInfo) {
-                                                  if (finished != NULL) {
-                                                    finished(imageContainer, error, downloadIdentifier, userInfo);
-                                                  }
-                                                }];
+    dispatch_queue_t callbackQueue = [self callbackQueue];
+    ASImageDownloaderProgress downloadProgress = NULL;
+    ASImageDownloaderCompletion completion = ^(id <ASImageContainerProtocol> _Nullable imageContainer, NSError * _Nullable error, id  _Nullable downloadIdentifier, id _Nullable userInfo) {
+      if (finished != NULL) {
+        finished(imageContainer, error, downloadIdentifier, userInfo);
+      }
+    };
+
+    if (self->_downloaderFlags.downloaderImplementsDownloadWithPriority
+        && self->_isInImageDownloaderPriorityExperiment) {
+      /*
+        Decide a priority based on the current interface state of this node.
+        It can happen that this method was called when the node entered preload state
+        but the interface state, at this point, tells us that the node is (going to be) visible.
+        If that's the case, we jump to a higher priority directly.
+       */
+      ASImageDownloaderPriority priority = ASImageDownloaderPriorityWithInterfaceState(interfaceState);
+
+      downloadIdentifier = [self->_downloader downloadImageWithURL:url
+                                                          priority:priority
+                                                     callbackQueue:callbackQueue
+                                                  downloadProgress:downloadProgress
+                                                        completion:completion];
+    } else {
+      /*
+        Kick off a download with default priority.
+        The actual "default" value is decided by the downloader.
+        ASBasicImageDownloader and ASPINRemoteImageDownloader both use ASImageDownloaderPriorityImminent
+        which is mapped to NSURLSessionTaskPriorityDefault.
+
+       This means that preload and display nodes use the same priority
+        and their requests are put into the same pool.
+      */
+      downloadIdentifier = [self->_downloader downloadImageWithURL:url
```

Comment:
same as above, if we can pass ASImageDownloaderPriorityImminent to priority, we can extract the call out of if else.